### ISSUE TITLE
***configure failed: arm-linux-androideabi-clang++ is not installed.

### ISSUE THREAD

**Issue by lordadamson** [kaldi-asr/kaldi#4123]
I was trying to cross-compile kaldi for android.

That's what I did:

```
# Set the PATH to contain paths to clang and arm-linux-androideabi-* utilities
export PATH=${NDK_BUNDLE_DIR}/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin:${NDK_BUNDLE_DIR}/toolchains/llvm/prebuilt/linux-x86_64/bin:$PATH

CXX=clang++ ./configure --static --use-cuda=no --android-incdir=${NDK_BUNDLE_DIR}/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include --host=arm-linux-androideabi --openblas-root=../../OpenBLAS/install
```

and this is the error I get:

```
Configuring KALDI to use OPENBLAS.
Backing up kaldi.mk to kaldi.mk.bak ...
Checking compiler arm-linux-androideabi-clang++ ...
***configure failed: arm-linux-androideabi-clang++ is not installed.
             You need g++ >= 4.7, Apple clang >= 5.0 or LLVM clang >= 3.3. ***
```

I check out the `toolchains/llvm/prebuilt/linux-x86_64/bin` directory. And indeed it does not have a `arm-linux-androideabi-clang++`.

So I work around that by creating a symlink to `armv7a-linux-androideabi30-clang++` like this: `ln -s armv7a-linux-androideabi30-clang++arm-linux-androideabi-clang++`

So maybe in previous versions of the NDK there was a file called `arm-linux-androideabi-clang++` but it doesn't exist now.

**Comment by kkm000** [comment]
Yup, the naming discrepancy is specific to arm-v7a, documented here: https://developer.android.com/ndk/guides/other_build_systems#overview

Let me see if I can fix that. I believe you can survive with the `ln -s` hack for a while? To the best of my knowledge, the NDK build support has been contributed by users outside the core team.

**Comment by lordadamson** [comment]
I noticed this part of the configure script:
```
  CXX="$HOST-$CXX"
  AR="$HOST-$AR"
  AS="$HOST-$AS"
  RANLIB="$HOST-$RANLIB"
```

And while it may be convenient to some users. I really dislike the fact that I am specifying a certain compiler, but the configure scripts overrides that with another.

I would find it much better if the user would explicitly specify the toolchain using the variables above.

So the configuration process would go something like this:

```
export PATH=${NDK_BUNDLE_DIR}/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin:${NDK_BUNDLE_DIR}/toolchains/llvm/prebuilt/linux-x86_64/bin:$PATH

export CXX=armv7a-linux-androideabi30-clang++
export AR=armv7a-linux-androideabi-ar
# optionally specify RANLIB and AS here

./configure --static --use-cuda=no --android-incdir=${NDK_BUNDLE_DIR}/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include --openblas-root=../../OpenBLAS/install
```

If you find that to be a good solution, I could create a pull request.

**Comment by danpovey** [comment]
Kirill, please decide about this, you know better than me about such things.

**Comment by kkm000** [comment]
@danpovey: Let me look into this. I rarely cross-compile, but my general fuzzy memory that specifying a "triple" is pretty much a standard approach (typically `--target=`, not `--host`, but that's ok).

@lordadamson: I believe you're on the right track, but then there are people out there who already configure their builds with the `--host`, so that a change would be rather breaking. What if we allow the environment variables as _overrides_ to values inferred from the triple? In your case, you'd need only `CXX=... ./configure --host=armv7a-linux-androideabi`, and those who do not need the overrides wouldn't have the configure command broken for them?

**Comment by lordadamson** [comment]
I don't think I understand. I will tell you what I understood, and you tell me what I got wrong.

The current implementation of `configure` prepends `HOST` to `CXX`, `AR`... etc.

So you're saying that we should do that, only if `CXX` was not defined by the user. Else -if it were defined by the user- we should use the `CXX` as is.

The statement above applies to the other tools in the toolchain as well.

Is that what you mean?

If that is so, then I'm afraid that will still break people's builds if they were using any other compiler that is not default.

**Comment by kkm000** [comment]
I do not really know how this part of code works. I'll look into its history.

Question: will your workaround currently do for you? Or are you blocked by this issue? I need to prioritize this, and we have only one of me...

**Comment by lordadamson** [comment]
Yeah, it's fine. My workaround works for now.

**Comment by kkm000** [comment]
Thanks, glad you have a workaround. So I'll put that not as the highest
priority issue on my list.

**Comment by kkm000** [comment]
@lordadamson, I looked at the code, and yes, CXX and friends are used in a simplistic way: if not set, then set to 'gcc', then if HOST set, make it `CXX=$HOST-$CXX`. So they do not work as overrides out of the box.

What I think will do, won't be breaking and easy to use at the same time is to apply transforms like this (using CXX w/o loss of generality):
1. Set `CXX=g++` if not set (this is the existing behavior).
2. If $CXX contains 3 or more dashes, leave it alone. This is new.
3. Otherwise, prepend `$HOST-` to it. This is also existing.

This way, people who use `--host=armv8-rpi3-linux-gnueabihf` (the GNU Raspberry Pi 3 toolchain) won't notice. Those who already set everything explicitly (like `AR=arm-linux-androideabi-ar` etc), won't either. But then you would be able to override only the compiler by doing
```
CXX=armv7a-linux-androideabi30-clang++ ./configure --host=arm-linux-androideabi ...
```
Easy enough to implement and requires less mucking with the triples¹, less potential for error. Do you see any downsides or non-covered use cases? Note that one can still explicitly specify every tool via the variables.

_________________
¹ I'm finding “triple” an amusing term for an identifier like `armv8-rpi3-linux-gnueabihf`. There are two hard problems in computer science: cache invalidation, naming things and off-by-one errors.

**Comment by lordadamson** [comment]
I think that's a pretty good solution. But I have two notes:

1-  The case when someone sets `CXX` to an absolute path where the compiler name doesn't follow the triple, it's instead expecting a `-target` or something. Ex: `CXX=/home/someone/toolchain/bin/clang++`

In that case it doesn't have the dashes but will have $HOST prepended to it. Maybe a solution would be to see if there are any slashes in `CXX` to see if this is a path.

2- The logic of the configure script will be very confusing to future maintainers. I guess it would help if some comments explained the decisions we took, and maybe a link to this issue.

**Comment by kkm000** [comment]
(1) makes sense. I chatted with someone running a small Kaldi-based business in the device niche, and they customize their build so much that this would be irrelevant. So all I have is a statistics of one point.

(2) not to worry, it's already convoluted enough it won't make it worse. We just want to cover basic cases, see (1) :)

**Comment by lordadamson** [comment]
Sounds good :+1:
